SPEC.md

```diff
@@ -334,32 +365,202 @@ interpreted as described in the following table:
+other characters in a string.
 
 ### Raw String
 
 Raw Strings in KDL are much like [Quoted Strings](#quoted-string), except they
 do not support `\`-escapes. They otherwise share the same properties as far as
-literal [Newline](#newline) characters go, and the requirement of UTF-8
-representation.
+literal [Newline](#newline) characters go, multi-line rules, and the requirement
+of UTF-8 representation.
 
-Raw String literals are represented as `r`, followed by zero or more `#`
-characters, followed by `"`, followed by any number of UTF-8 literals. The
-string is then closed by a `"` followed by a _matching_ number of `#`
-characters. This allows them to contain raw `"` or `#` characters; only the
-precise terminator (resembling `"##`, for example) ends the raw string. This
-means that the string sequence `"` or `"#` and such must not match the closing
-`"` with the same or more `#` characters as the opening `r`.
+Raw String literals are represented with one or more `#` characters, followed
+by `"`, followed by any number of UTF-8 literals. The string is then closed by
+a `"` followed by a _matching_ number of `#` characters. This means that the
+string sequence `"` or `"#` and such must not match the closing `"` with the
+same or more `#` characters as the opening `#`, in the body of the string.
+
+Like other Strings, Raw Strings _MUST NOT_ include any of the [disallowed
+literal code-points](#disallowed-literal-code-points) as code points in their
+body. Unlike with Quoted Strings, these cannot simply be escaped, and are thus
+unrepresentable when using Raw Strings.
 
 #### Example
 
 ```kdl
-just-escapes r"\n will be literal"
-quotes-and-escapes r#"hello\n\r\asd"world"#
+just-escapes #"\n will be literal"#
+```
+
+The string contains the literal characters `\n will be literal`.
+
+```kdl
+quotes-and-escapes ##"hello\n\r\asd"#world"##
+```
+
+The string contains the literal characters `hello\n\r\asd"#world`
+
+
+### Multi-line Strings
+
+When a Quoted or Raw String spans multiple lines with literal, non-escaped
+Newlines, it follows a special multi-line syntax that automatically "dedents"
+the string, allowing its value to be indented to a visually matching level if
+desired.
+
+A Multi-line string _MUST_ start with a [Newline](#newline) immediately
+following its opening `"`. Its final line _MUST_ contain only whitespace,
+followed by a single closing `"`. All in-between lines that contain
+non-whitespace characters _MUST_ start with the exact same whitespace as the
+final line (precisely matching codepoints, not merely counting characters).
+
+The value of the Multi-line String omits the first and last Newline, the
+Whitespace of the last line, and the matching Whitespace prefix on all
+intermediate lines. The first and last Newline can be the same character (that
+is, empty multi-line strings are legal).
+
+Strings with literal Newlines that do not immediately start with a Newline and
```

Comment:
Should this be non-escaped literal newlines?

Reply:
oh yes.

So:

```kdl
foo "\
string"
```

should be a "single-line" string, yes?

Reply:
actually do we even allow this? Do we want to?...

Reply:
It is allowed in the grammar, it seems.

Reply:
We don't currently allow that; if you have a multiline, the closing quote must be on a line of its own, with only the dedent space accompanying it.

I don't think we want to allow it; it complicates the model a bit (right now, it's clear that the string content is strictly the lines between the first and last), and since it wouldn't interact with dedent, your string is by necessity already starting from the left edge of your document, so having the closing quote down there as well is fine, visually.

Reply:
Yeah, it's allowed, and I think it's acceptable, if weird. A string is considered "multi-line", per the spec, when it contains one or more *non-escaped* newlines.

Reply:
Oh yes. No problems, then.